### ISSUE TITLE
[BUG]: Data leakage in pycaret 3 classification with unbalanced dataset?

### ISSUE THREAD

**Issue by sorenwacker** [pycaret/pycaret#3507]
### pycaret version checks

- [X] I have checked that this issue has not already been reported [here](https://github.com/pycaret/pycaret/issues).

- [X] I have confirmed this bug exists on the [latest version](https://github.com/pycaret/pycaret/releases) of pycaret.

- [X] I have confirmed this bug exists on the master branch of pycaret (pip install -U git+https://github.com/pycaret/pycaret.git@master).

### Issue Description

I am getting a test AUC of 1 with tree based models with random data. 

models tested rt, xgboost, lightgbm, et, dt

### Reproducible Example

```python
from pycaret.classification import *

# Import libraries
import numpy as np
import pandas as pd

# Set random seed for reproducibility
np.random.seed(42)

# Define number of samples
N = 30000

# Generate features
numeric_feature = np.random.normal(0, 1, N)
cat_feature = np.random.choice(31, N).astype(int)
cat_feature2 = np.random.choice(2, N).astype(str)

# Combine features into a single matrix
X = np.column_stack((cat_feature, cat_feature2))

# Generate target variable
y = np.random.binomial(1, 0.2, N)

# Convert to DataFrame and add target variable
df = pd.DataFrame(X, columns=['categorical-cardinal', 'categorical-binary'])
df['target'] = y

# Print the first five rows of the dataset
print("Synthetic Dataset for Machine Learning:\n")
print(df.head())

setup(df, target='target', numeric_features=[], categorical_features=['categorical-cardinal', 'categorical-binary'])

model = create_model('lightgbm')

plot_model(model, 'auc')

predict_model(model)
```

### Expected Behavior

The test auc should be around 0.5, as in pycaret version 2.

![image](https://user-images.githubusercontent.com/3391614/234414110-61eb88e9-5920-4516-b0eb-bf9c92af8236.png)

### Actual Results

![image](https://user-images.githubusercontent.com/3391614/234413279-d5404116-76ef-4ff5-ba1e-54deb339f4e3.png)

```python-traceback
No error message, but unrealistic results.
```

### Installed Versions

<details>
System:
    python: 3.10.10 (main, Mar 21 2023, 18:45:11) [GCC 11.2.0]
executable: /bulk/LSARP/envs/conda/pycaret3/bin/python
   machine: Linux-4.18.0-425.10.1.el8_7.x86_64-x86_64-with-glibc2.31

PyCaret required dependencies:
                 pip: 23.0.1
          setuptools: 66.0.0
             pycaret: 2.1.post14082020
             IPython: 7.34.0
          ipywidgets: 7.7.5
                tqdm: 4.64.1
               numpy: 1.23.5
              pandas: 1.5.3
              jinja2: 3.1.2
               scipy: 1.9.3
              joblib: 1.2.0
             sklearn: 1.2.2
                pyod: 1.0.9
            imblearn: 0.10.1
   category_encoders: 2.6.0
            lightgbm: 3.3.5
               numba: 0.56.4
            requests: 2.28.2
          matplotlib: 3.6.3
          scikitplot: 0.3.7
         yellowbrick: 1.5
              plotly: 5.14.1
             kaleido: 0.2.1
         statsmodels: 0.13.5
              sktime: 0.17.0
               tbats: 1.1.3
            pmdarima: 2.0.3
              psutil: 5.9.5

PyCaret optional dependencies:
                shap: 0.41.0
           interpret: 0.3.2
                umap: 0.5.3
    pandas_profiling: 3.6.6
  explainerdashboard: 0.4.2.1
             autoviz: 0.1.601
           fairlearn: 0.7.0
             xgboost: 1.7.5
            catboost: 1.1.1
              kmodes: 0.12.2
             mlxtend: 0.22.0
       statsforecast: 1.5.0
        tune_sklearn: 0.4.5
                 ray: 2.3.1
            hyperopt: 0.2.7
              optuna: 3.1.1
               skopt: 0.9.0
              mlflow: 1.30.1
              gradio: 3.27.0
             fastapi: 0.95.1
             uvicorn: 0.21.1
              m2cgen: 0.10.0
           evidently: 0.3.0
               fugue: 0.8.3
           streamlit: Not installed
             prophet: Not installed
​</details>

**Comment by tvdboom** [comment]
I believe the issue is with your dummy dataset. You are creating a dataset with 2 features with a maximum of 62 possible combinations. On 30k rows, that doesn't make sense. If you increase the number of possible combinations, e.g.

```
# Generate features
numeric_feature = np.random.normal(0, 1, N)
cat_feature = np.random.choice(310000, N).astype(int)
cat_feature2 = np.random.choice(2, N).astype(str)
```

the results are as expected. Also with other datasets, like the titanice one, the results are as expected.

**Comment by ngupta23** [comment]
@tvdboom What is odd though is that with the original example provided, the training data has a good mix of target = 1 and target = 0 for the same combination of features. The same is true for the test dataset. So the best case scenario here would be that the model picks either 0 or 1 as the prediction for a combination of features, but should not switch between these a prediction of 0 and 1 in the test set.

I took one example below. Not sure how the model would come to the conclusion that one test observation with `categorical-cardinal == '17' and categorical-binary == '0'` should be predicted as 1 and another test observation with the same features should be predicted as 0. Possibly the model has access to the labels somehow?


![image](https://user-images.githubusercontent.com/33585645/236624994-2c0bd9b8-1d79-4576-93c1-e0587be4181e.png)

**Comment by tvdboom** [comment]
that's odd indeed. I'll keep digging

**Comment by tvdboom** [comment]
Alright, I understand it now. 

What's happening is that the encoder is encoding every combination wrt the target with a unique value, hence the perfect prediction. For example:
![image](https://user-images.githubusercontent.com/32366550/236642684-c534c495-e988-4f42-83e6-e895e026bcca.png)

This happens because LeaveOneOutEncoder takes the mean of the target values minus the target value of the row you are calculating. Therefore, if the target value is 1, the result is different than if it's 0.

If instead, you would use TargetEncoder like this:

```python
from category_encoders.target_encoder import TargetEncoder

setup(df, target='target', categorical_features=['categorical-cardinal', 'categorical-binary'], encoding_method=TargetEncoder())
```

You'll see that the results are as expected, because the encoded values are no longer unique.

**Comment by tvdboom** [comment]
I found the issue.

The encoding of the `categorical-cardinal` column is different depending on the target value, hence the perfect score. See rows 1, 2 and 4.
![image](https://user-images.githubusercontent.com/32366550/236647821-ca3c69d2-624d-45dd-bdda-d82926266907.png)
![image](https://user-images.githubusercontent.com/32366550/236647791-e1298d75-60a4-48a0-834f-110a147e8216.png)

This happens because [LeaveOneOutEncoder](https://contrib.scikit-learn.org/category_encoders/leaveoneout.html) behaves differently if transforming when `y` is provided and when it's not. When transforming on new data, `y` should not be provided, but in pycaret's implementation of `X_test_transformed`, `y` is provided. We need it there though because of other transformers that could require it, so we can not just remove it. I'll think of a solution.

In the mean time, using another encoder like TargetEncoder works as expected.
```python
from category_encoders.target_encoder import TargetEncoder

setup(df, target='target', numeric_features=[], categorical_features=['categorical-cardinal', 'categorical-binary'], encoding_method=TargetEncoder())
```

**Comment by ngupta23** [comment]
```
pycaret's implementation of X_test_transformed, y is provided
```

In time series, we separate out y into its own y_test_transformed. So X_test_transformed just has the X values and no y values. Then when we need to pass y to a transformer using test data, we have more control over it. Maybe a similar approach will help.

**Comment by tvdboom** [comment]
We can't do that. `X` and `y` need to be transformed together since there are transformers that drop rows. I think the solution would be to change `LeaveOneOut` for `TargetEncoder` as default encoder and avoid the use of LeaveOneOut since it does break with sklearn's API. See this issue https://github.com/scikit-learn-contrib/category_encoders/issues/167#issuecomment-498305669
